node/common/src/service/archive.rs

```diff
@@ -0,0 +1,490 @@
+    }
+
+    fn run(
+        mut archive_receiver: mpsc::UnboundedReceiver<ArchiveTransitionFronntierDiff>,
+        address: SocketAddr,
+    ) {
+        while let Some(breadcrumb) = archive_receiver.blocking_recv() {
+            let mut retries = ARCHIVE_SEND_RETRIES;
+            while retries > 0 {
+                match rpc::send_diff(address, v2::ArchiveRpc::SendDiff(breadcrumb.clone())) {
+                    Ok(result) => {
+                        if result.should_retry() {
+                            node::core::warn!(node::core::log::system_time(); summary = "Archive suddenly closed connection, retrying...");
+                            retries -= 1;
+                            std::thread::sleep(std::time::Duration::from_millis(RETRY_INTERVAL_MS));
+                        } else {
+                            node::core::warn!(node::core::log::system_time(); summary = "Successfully sent diff to archive");
+                            break;
+                        }
+                    }
+                    Err(e) => {
+                        node::core::warn!(node::core::log::system_time(); summary = "Failed sending diff to archive", error = e.to_string(), retries = retries);
+                        retries -= 1;
+                        std::thread::sleep(std::time::Duration::from_millis(RETRY_INTERVAL_MS));
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn start(address: SocketAddr) -> Self {
+        let (archive_sender, archive_receiver) =
+            mpsc::unbounded_channel::<ArchiveTransitionFronntierDiff>();
+
+        thread::Builder::new()
+            .name("openmina_archive".to_owned())
+            .spawn(move || {
+                Self::run(archive_receiver, address);
+            })
+            .unwrap();
+
+        Self::new(archive_sender)
+    }
+}
+
+impl node::transition_frontier::archive::archive_service::ArchiveService for NodeService {
+    fn send_to_archive(&mut self, data: ArchiveTransitionFronntierDiff) {
+        if let Some(archive) = self.archive.as_mut() {
+            if let Err(e) = archive.archive_sender.send(data.clone()) {
+                node::core::warn!(
+                    node::core::log::system_time();
+                    summary = "Failed sending diff to archive service",
+                    error = e.to_string()
+                );
+            }
+        }
+    }
+}
+
+// We need to replicate the ocaml node's RPC like interface
+mod rpc {
+    use binprot::BinProtWrite;
+    use mina_p2p_messages::rpc_kernel::{Message, NeedsLength, Query, RpcMethod};
+    use mina_p2p_messages::v2::{self, ArchiveRpc};
+    use mio::event::Event;
+    use mio::net::TcpStream;
+    use mio::{Events, Interest, Poll, Registry, Token};
+    use std::io::{self, Read, Write};
+    use std::net::SocketAddr;
+
+    const MAX_RECURSION_DEPTH: u8 = 25;
+
+    // messages
+    const HEADER_MSG: [u8; 7] = [2, 253, 82, 80, 67, 0, 1];
```

Comment:
probably not worth the effort to change this now if it is already working, but are the raw bytes really needed? iirc we had already found the source of this in the OCaml code.